### PR TITLE
Better handling of databse connections errors

### PR DESCRIPTION
**Pull request by iNecas**
1. Retry several times when db error occurs
2. Terminate when the retries don't help
3. Let the executor to finish just after the workers are done:
   don't wait for the whole task to finish.

### CODE REVIEW

File: lib/dynflow/errors.rb [hunk 1]
```diff
@@ -24,5 +24,13 @@ def inspect
...
       end
     end
+
+    class PersistenceError < StandardError
+      def self.delegate(original_exception)
+        self.new("caused by #{original_exception.class}: #{original_exception.message}").tap do |e|
+          e.set_backtrace original_exception.backtrace
+        end
+      end
+    end
```
Comment (pitr-ch) [inline lib/dynflow/errors.rb]: what about `justified` gem?

File: lib/dynflow/world.rb [hunk 2]
```diff
@@ -32,7 +33,7 @@ def default_options
     end
 
     def clock
-      @clock ||= Clock.new(logger)
+      @clock_barrier.synchronize { @clock ||= Clock.new(logger) }
```
Comment (pitr-ch) [inline lib/dynflow/world.rb]: Could we set up the clock in the `initialize` method? Then we can avoid the barrier.

File: lib/dynflow/errors.rb (hunk 1 above)
Comment (iNecas) [inline lib/dynflow/errors.rb]: Adding additional dependency is not justified for this simple code

File: lib/dynflow/world.rb (hunk 2 above)
Comment (iNecas) [inline lib/dynflow/world.rb]: I was thinking about it, but in the remote executor case, the clock is seldom used: so we will have additional thread in every client instance.

**Comment by iNecas** [comment]
Resolved issues with event futures not getting resolved occasionally on termination

File: lib/dynflow/errors.rb (hunk 1 above)
Comment (pitr-ch) [inline lib/dynflow/errors.rb]: ok :D

File: lib/dynflow/world.rb (hunk 2 above)
Comment (pitr-ch) [inline lib/dynflow/world.rb]: hm its sleeping thread vs synchronizing all the time. Probably not important.

**Comment by pitr-ch** [comment]
ack

**Comment by iNecas** [comment]
I've two more commits for exiting the process when terminating and handling an edge case of some plan being already in paused state when finishing the termination

**Comment by pitr-ch** [comment]
:+1:

**Comment by iNecas** [comment]
@pitr-ch one last update: I added the termination when the db connection fails constantly only when happens inside the executor: the client part doens't necessary need to fail when this occurs.

**Comment by pitr-ch** [comment]
LGTM

**Pull request merged**